File: pomade-signer-go/schema.go
package main

type Commit struct {
	Idx      uint32 `json:"idx"`
	Pubkey   string `json:"pubkey"`
	HiddenPn string `json:"hidden_pn"`
	BinderPn string `json:"binder_pn"`
}

type Group struct {
	Commits   []Commit `json:"commits"`
	GroupPk   string   `json:"group_pk"`
	Threshold uint32   `json:"threshold"`
}

type Share struct {
	Idx      uint32 `json:"idx"`
	BinderSn string `json:"binder_sn"`
	HiddenSn string `json:"hidden_sn"`
	Seckey   string `json:"seckey"`
}

type SessionItem struct {
	Pubkey       string  `json:"pubkey"`
	Client       string  `json:"client"`
	CreatedAt    uint64  `json:"created_at"`
	Deactivated  *uint64 `json:"deactivated_at,omitempty"`
	LastActivity uint64  `json:"last_activity"`
	Threshold    uint32  `json:"threshold"`
	Total        uint32  `json:"total"`
	Idx          uint32  `json:"idx"`
	Email        *string `json:"email,omitempty"`
}

type AuthPayload struct {
	EmailHash    string `json:"email_hash"`
	PasswordHash string `json:"password_hash,omitempty"`
	OTP          string `json:"otp,omitempty"`
}

func (a AuthPayload) IsPassword() bool {
	return a.PasswordHash != ""
}

func (a AuthPayload) IsOTP() bool {
	return a.OTP != ""
}

type RegisterRequest struct {
	Share    Share `json:"share"`
	Group    Group `json:"group"`
	Recovery bool  `json:"recovery"`
}

type RegisterResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
}

type SignRequestInner struct {
	Content *string    `json:"content"`
	Hashes  [][]string `json:"hashes"`
	Members []uint32   `json:"members"`
	Stamp   uint64     `json:"stamp"`
	Type    string     `json:"type"`
	Gid     string     `json:"gid"`
	Sid     string     `json:"sid"`
}

type SignRequest struct {
	Request SignRequestInner `json:"request"`
}

type SignResult struct {
	Idx    uint32      `json:"idx"`
	Psigs  [][2]string `json:"psigs"`
	Pubkey string      `json:"pubkey"`
	Sid    string      `json:"sid"`
}

type SignResponse struct {
	OK      bool        `json:"ok"`
	Message string      `json:"message"`
	Result  *SignResult `json:"result,omitempty"`
}

type SignCommitRequest struct {
	Members []uint32 `json:"members"`
}

type SignCommitResult struct {
	CommitID string `json:"commit_id"`
	Idx      uint32 `json:"idx"`
	Pubkey   string `json:"pubkey"`
	HiddenPn string `json:"hidden_pn"`
	BinderPn string `json:"binder_pn"`
}

type SignCommitResponse struct {
	OK      bool              `json:"ok"`
	Message string            `json:"message"`
	Result  *SignCommitResult `json:"result,omitempty"`
}

type PublicNonceItem struct {
	Idx      uint32 `json:"idx"`
	HiddenPn string `json:"hidden_pn"`
	BinderPn string `json:"binder_pn"`
}

type SignCompleteRequestInner struct {
	Content *string  `json:"content"`
	Hash    []string `json:"hash"`
	Members []uint32 `json:"members"`
	Stamp   uint64   `json:"stamp"`
	Type    string   `json:"type"`
	Gid     string   `json:"gid"`
	Sid     string   `json:"sid"`
}

type SignCompleteRequest struct {
	CommitID string                   `json:"commit_id"`
	Request  SignCompleteRequestInner `json:"request"`
	Pnonces  []PublicNonceItem        `json:"pnonces"`
}

type SignCompleteResult struct {
	Idx    uint32    `json:"idx"`
	Psig   [2]string `json:"psig"`
	Pubkey string    `json:"pubkey"`
	Sid    string    `json:"sid"`
}

type SignCompleteResponse struct {
	OK      bool                `json:"ok"`
	Message string              `json:"message"`
	Result  *SignCompleteResult `json:"result,omitempty"`
}

type EcdhRequest struct {
	Idx     uint32   `json:"idx"`
	Members []uint32 `json:"members"`
	EcdhPk  string   `json:"ecdh_pk"`
}

type EcdhResult struct {
	Idx      uint32   `json:"idx"`
	Keyshare string   `json:"keyshare"`
	Members  []uint32 `json:"members"`
	EcdhPk   string   `json:"ecdh_pk"`
}

type EcdhResponse struct {
	OK      bool        `json:"ok"`
	Message string      `json:"message"`
	Result  *EcdhResult `json:"result,omitempty"`
}

type RecoverySetupRequest struct {
	Email        string `json:"email"`
	PasswordHash string `json:"password_hash"`
}

type RecoverySetupResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
}

type ChallengeRequest struct {
	Prefix    string `json:"prefix"`
	EmailHash string `json:"email_hash"`
}

type ChallengeResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
}

type LoginStartRequest struct {
	Auth AuthPayload `json:"auth"`
}

type LoginStartResponse struct {
	OK      bool          `json:"ok"`
	Message string        `json:"message"`
	Items   []SessionItem `json:"items,omitempty"`
}

type LoginSelectRequest struct {
	Client string `json:"client"`
}

type LoginSelectResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
	Group   *Group `json:"group,omitempty"`
}

type RecoveryStartRequest struct {
	Auth AuthPayload `json:"auth"`
}

type RecoveryStartResponse struct {
	OK      bool          `json:"ok"`
	Message string        `json:"message"`
	Items   []SessionItem `json:"items,omitempty"`
}

type RecoverySelectRequest struct {
	Client string `json:"client"`
}

type RecoverySelectResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
	Share   *Share `json:"share,omitempty"`
	Group   *Group `json:"group,omitempty"`
}

type SessionListResponse struct {
	OK      bool          `json:"ok"`
	Message string        `json:"message"`
	Items   []SessionItem `json:"items"`
}

type SessionListRequest struct{}

type SessionDeactivateRequest struct {
	Client string `json:"client"`
}

type SessionDeactivateResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
}

type SessionDeleteRequest struct {
	Client string `json:"client"`
}

type SessionDeleteResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
}

File: packages/core/src/schema.ts
import * as z from "zod"

// Security limits to prevent DoS attacks via unbounded payloads
const MAX_HASHES_PER_REQUEST = 10 // Maximum number of hashes in a single signature request
const MAX_MEMBERS = 5 // Maximum number of members in a signing group
const MAX_COMMITS = 5 // Maximum number of commits in a group

const hex = z
  .string()
  .regex(/^[0-9a-fA-F]*$/)
  .refine(e => e.length % 2 === 0)
const hex32 = hex.refine(e => e.length === 64)
const hex33 = hex.refine(e => e.length === 66)

const commit = z.object({
  idx: z.number(),
  pubkey: hex33,
  hidden_pn: hex33,
  binder_pn: hex33,
})

const publicNonceItem = z.object({
  idx: z.number(),
  hidden_pn: hex33,
  binder_pn: hex33,
})

const group = z.object({
  commits: z.array(commit).max(MAX_COMMITS),
  group_pk: hex33,
  threshold: z.number(),
})

const share = z.object({
  idx: z.number(),
  seckey: hex32,
})

const psig_entry = z.tuple([hex32, hex32])

// Use tuple with rest to maintain type compatibility while enforcing max length
const sighash_vec = z
  .tuple([hex32])
  .rest(hex32)
  .refine(arr => arr.length <= MAX_HASHES_PER_REQUEST, {
    message: `Maximum ${MAX_HASHES_PER_REQUEST} hashes allowed per request`,
  })

const sessionItem = z.object({
  pubkey: hex32,
  client: hex32,
  created_at: z.int().positive(),
  deactivated_at: z.int().optional(),
  last_activity: z.int().positive(),
  threshold: z.int().positive(),
  total: z.number(),
  idx: z.number(),
  email: z.string().email().optional(),
})

const passwordAuth = z.object({
  email_hash: z.string(),
  password_hash: z.string(),
})

const otpAuth = z.object({
  email_hash: z.string(),
  otp: z.string(),
})

const auth = z.union([passwordAuth, otpAuth])

export type SessionItem = z.infer<typeof sessionItem>
export type PasswordAuth = z.infer<typeof passwordAuth>
export type OTPAuth = z.infer<typeof otpAuth>
export type Auth = z.infer<typeof auth>

export const isPasswordAuth = (auth: Auth): auth is PasswordAuth =>
  Boolean((auth as unknown as PasswordAuth).password_hash)

export const isOTPAuth = (auth: Auth): auth is OTPAuth => Boolean((auth as unknown as OTPAuth).otp)

export const Schema = {
  registerRequest: z.object({
    share: share,
    group: group,
    recovery: z.boolean(),
  }),
  registerResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
  }),
  signCommitRequest: z.object({
    members: z.number().array().max(MAX_MEMBERS),
  }),
  signCommitResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    result: z
      .object({
        commit_id: hex32,
        idx: z.number(),
        pubkey: hex33,
        hidden_pn: hex33,
        binder_pn: hex33,
      })
      .optional(),
  }),
  signCompleteRequest: z.object({
    commit_id: hex32,
    request: z.object({
      content: z.string().nullable(),
      hash: sighash_vec,
      members: z.number().array().max(MAX_MEMBERS),
      stamp: z.number(),
      type: z.string(),
      gid: hex32,
      sid: hex32,
    }),
    pnonces: publicNonceItem.array().max(MAX_MEMBERS),
  }),
  signCompleteResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    result: z
      .object({
        idx: z.number(),
        psig: psig_entry,
        pubkey: hex33,
        sid: hex32,
      })
      .optional(),
  }),
  ecdhRequest: z.object({
    idx: z.number(),
    members: z.number().array().max(MAX_MEMBERS),
    ecdh_pk: hex32,
  }),
  ecdhResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    result: z
      .object({
        idx: z.number(),
        keyshare: hex,
        members: z.number().array().max(MAX_MEMBERS),
        ecdh_pk: hex,
      })
      .optional(),
  }),
  recoverySetupRequest: z.object({
    email: z.string().email(),
    password_hash: z.string(),
  }),
  recoverySetupResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
  }),
  challengeRequest: z.object({
    prefix: z.string().regex(/^\d{2}$/),
    email_hash: z.string(),
  }),
  challengeResponse: z.object({
    ok: z.literal(true),
    message: z.string(),
  }),
  loginStartRequest: z.object({
    auth,
  }),
  loginStartResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    items: z.array(sessionItem).optional(),
  }),
  loginSelectRequest: z.object({
    client: hex32,
  }),
  loginSelectResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    group: group.optional(),
  }),
  recoveryStartRequest: z.object({
    auth,
  }),
  recoveryStartResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    items: z.array(sessionItem).optional(),
  }),
  recoverySelectRequest: z.object({
    client: hex32,
  }),
  recoverySelectResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    share: share.optional(),
    group: group.optional(),
  }),
  sessionListRequest: z.object({}),
  sessionListResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
    items: z.array(sessionItem).optional(),
  }),
  sessionDeactivateRequest: z.object({
    client: hex32,
  }),
  sessionDeactivateResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
  }),
  sessionDeleteRequest: z.object({
    client: hex32,
  }),
  sessionDeleteResponse: z.object({
    ok: z.boolean(),
    message: z.string(),
  }),
}
